[MRG] Allow sample weights and other fit_params for RFE

g-rutter opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

Fixes  #7308
#### What does this implement/fix? Explain your changes.

Adds support for passing sample weights into `RFE.fit()` and having them used by the estimator's `fit` method.

Adds a test for this with the iris dataset, where sample weights are used to give one class double its normal weight and compare that to doubling the samples in that class. The test passes if both these approaches produce the same feature ranking. This feature ranking is different from the one which arises when all samples have the same weight.
#### Any other comments?

I will look in to adding the same support for `RFECV`, and to `RFE.score()`.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Let RFE pass fit_params to estimator

Commit:
Test for rfe sample weights

Commit:
Switch from np.full to np.ones for compatibility. Make the class being tested a variable, .

amueller commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -75,6 +75,37 @@ def test_rfe_features_importance():
+    rfe.fit(X, y, sample_weight=w)
+    ranking_1 = rfe.ranking_.copy()
+
+    # Case 2 - duplicate the features of one class
```
duplicate the samples?

g-rutter replied:
Agreed, I'll change this.

amueller commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -75,6 +75,37 @@ def test_rfe_features_importance():
+    rfe.fit(X2, y2)
+    ranking_2 = rfe.ranking_.copy()
+
+    assert_array_equal(ranking_1, ranking_2)
```
did either of them actually change the result? Maybe also test that they are not the same as without weighting?

g-rutter replied:
Yes, this changes the result from the unaltered case. I can add a test that this is indeed the case, just to help a viewer understand.

amueller commented:
Looks good apart from comment about the test.

g-rutter commented:
I do consider this a work in progress though, due to the remaining additions mentioned at the end of the first post.

amueller commented:
``score`` doesn't take ``sample_weights`` (for now)

Commit:
Improve sample weights testing

- Test that the weighted feature ranking is different from the original
  feature ranking
- Clearer comments and variable names

g-rutter commented:
I have updated the tests with respect to @amueller's comments, but won't work on `RFECV` until the interaction of this pull request with #1574 is clarified. In fact, these two pull requests seem to have overlapping purposes. However, my pull request offers full `fit_params`, whereas #1574 offers only sample weights, for `RFE`.

Commit:
Fix nosetests for old versions of numpy by not using  as a context manager.

miguellacerda commented:
I think you also need to add the `**fit_params` argument to the fit method called just after line 200 (near the end of `_fit`):
```
# Set final attributes
features = np.arange(n_features)[support_]
self.estimator_ = clone(self.estimator)
self.estimator_.fit(X[:, features], y, **fit_params)
```

lschneidpro commented:
any updates ?

amueller commented:
This needs a merge with master

hakeemo commented:
any updates ?

proinsias commented:
@g-rutter – any updates on this?

cerlymarco commented:
Everyone looking for RFE with Gradient Boosting, like LGBM or XGB, I suggest [shap-hypetune](https://github.com/cerlymarco/shap-hypetune)... _A python package for simultaneous Hyperparameters Tuning and Features Selection for Gradient Boosting Models_. 

It supports RFE (also with shap feature ranking) with every fitting parameters like in the standard algorithm API

fbidu commented:
Me and @ijpulidos were taking over this PR to finish it up against the current main branch
